### ISSUE TITLE
A.01.validate.md

### ISSUE THREAD

**Issue by aquaglia** [inspire-eu-validation/ats-metadata#10]
The description of the ATC says:
> Document shall pass schema validation without errors, using one of the following XML schema definitions:
* schemaLocation A
* schemaLocation B
* schemaLocation C

We need to specify the criteria to determine which schemaLocation to use.
Otherwise a validator, in order to claim compliance, would have to try A and if it fails B and if it fails C.

**Comment by pvgenuchten** [comment]
@aquaglia, what about the case that an xsd is duplicated and referenced locally (on the data server). In this case the schema is equal, but the location is different

And what about the case if the Schema is extended. In that case the schema's are different but could still be interoperable?

**Comment by aquaglia** [comment]
@pvgenuchten, where a specific schema definition is physically located is irrelevant.
The schema definitions in themselves are relevant.
In particular, the schema definitions found at the second and third schema locations in the ATC description are semantically different, and I am referring to:
ISO 19139 version 2005-DIS with GML 3.2.0
ISO/TS 19139:2007 with GML 3.2.1:
You need to specify an algorithm through which a reader can determine against which of the two schema definitions, the validation is intended to succeed.

**Comment by ilkkarinne** [comment]
As I'm sure you now, generally, the namespace of the XML Schema should be enough to identify the schema for validation purposes. The schemaLocation attribute is provided to help the validators in fetching the schema rules if it does not already know the referred schema: it's a hint.

The author of an XML document should not assume that the reader will necessarily use the schemaLocation provided to fetch the schema file. It may already have fetched the same schema (=same namespace) for interpreting some previously processed XML documents, or it may be designed to use locally cached versions of the schema files for performance or data security reasons (these files are also identified by namespaces, not schemaLocation URLs).

Unfortunately the case of the gmd is a good example what happens when this principle is not followed, but instead different versions of the schema files with the same namespace are published in different locations and by different parties. The identification of the schema rules is no longer possible using the namespace alone, which creates big problems for the validators. I agreed with @aquaglia that for this particular case we should suggest a way for the data providers to declare and validators to identify which variant of the gmd schema is used for a particular metadata record.

As for the @pvgenuchten's question of extended schemas: If these "extended" schemas are different from the original schemas they should be published under a different namespace. This in turn makes them completely different from the validator's perspective, because even the elements and types left unchanged are now contained in a different namespace, and thus are not equal.

**Comment by pvgenuchten** [comment]
@ilkkarinne, it might be the case that in general a namespace is enough to identify a schema, this is not the case in iso-metadata. We have quite some examples of namespace-naming being kept equal, while the underlying schema changed.

I think at this point we can choose some approaches:
- Develop an algorythm as @aquaglia suggests that will grab the xsd and check if it's similar as the appropriate schema
- Assume some of the other tests in this suite will likely fail if the appropriate schema is not provided, so just check if the document validates against the schema it's advertising to validate against and continue with the other tests
- As approach 2, but additionally, replace the customised schema locations by the appropriate schema locations and validate the document again.

@ilkkarinne your comment 'makes them completely different from the validator's perspective' suggests that the validator should not validate a metadata document that is based on an extended schema.

**Comment by aquaglia** [comment]
@pvgenuchten In no way did I intend to suggest a schema comparison.
I meant instead detecting elements in the document instance that can drive the selection of the schema implementation to use.

**Comment by pvgenuchten** [comment]
@aquaglia ah I see it now, in that case the test will:
- grab the xsd
- algorythm to test if it's schema "ISO 19139 version 2005-DIS with GML 3.2.0" or "ISO/TS 19139:2007 with GML 3.2.1:"
- Validate against one of these

**Comment by aquaglia** [comment]
@pvgenuchten 
No, no, you have to inspect the xml document instance to validate not the resource pointed by schemaLocation

The schemaLocation attribute is not mandatory and we shall not rely on it.

**Comment by ilkkarinne** [comment]
@aquaglia : do you mean that the validator should decide which schema variation the metadata record is using by checking for the existence of certain elements in the record and only then run the schema validation against that schema? This option would also mean that this "peeking" of elements should be done using a non-validating XML parser.

**Comment by aquaglia** [comment]
Yes, of course.
It is a two step process:
1) You detect the "type" of the xml document, so that you know which schema definition files you have to pick;
2) You validate the xml document.

**Comment by pvgenuchten** [comment]
@aquaglia you're suggesting that implementors should not extend the schema or use a compatible schema. Is this only for metadata, or do you consider this the case for all test suites, what if people extend the WMS capabilities schema or a data theme schema, which seems to be encouraged to enable inclusion of organisation specific attributes.

I know schemaLocation attribute is (unfortunately) not mandatory, but I guess if it is specified it would make sense to use it in a test.

**Comment by aquaglia** [comment]
@pvgenuchten Each XML schema has its own rules for handling extensions and customizations. It is not possible to generalize.

The point of validating an XML document is to ensure its syntactic structure is consistent with one that the consumer of the document knows and understands.

It really does not help the consumer to know or check that an XML document is valid with respect to a schema modified by the producer of the document.
It also does not make sense to ask the consumer to embark on a semantic evaluation of the schemas customized in arbitrary ways by the producer.

If your proposal is to amend the MD TGs in order to make schemaLocation mandatory for metadata, I am against it especially if you also intend to allow the schemaLocation to point to customized versions of the schemas and perform the validation using those schemas.

@michellutz FYI

**Comment by PeterParslow** [comment]
And remember that:
1, the official schemas for ISO 19139:2007 (at http://standards.iso.org/iso/19139/) do not include metadata for services. So far as I know, the nearest-to-official set was prepared for a late draft of 19119, and is at http://schemas.opengis.net/iso/19139/20060504/.

2. neither does ISO 19139:2007 include metadata 'extensions for imagery', but at least that (19115-2) does have an official ISO schema (http://standards.iso.org/iso/19115/-2/gmi/1.0/)

3. some times it may be better to start at the gmx.xsd root - although the only specific I know is if you are using a 'hybrid' schema set that contains service metadata as well as dataset metadata.

(I agree that the 'physical' schema locations - URLs - aren't the official identifiers, I'm using them as convenient indicators of the mess that OGC/ISO had created before agreeing that namespace identifiers need to be versioned)
